### PR TITLE
Add server resource pack API

### PR DESCRIPTION
**Pull request by yusshu**
This would allow plugins to set the server resource pack so it is sent by the server

### CODE REVIEW

File: patches/server/0803-Implement-server-resource-pack-API.patch [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+         return datapackManager;
+     }
+ 
++    @Override
++    public void setResourcePack(@Nonnull String url, @Nonnull String hash, boolean required, @Nullable net.kyori.adventure.text.Component prompt) {
++        Validate.notNull(url, "Resource pack URL cannot be null");
++        Validate.notNull(hash, "Hash cannot be null");
++        net.minecraft.network.chat.Component promptComponent = prompt != null ?
++            io.papermc.paper.adventure.PaperAdventure.asVanilla(prompt) :
++            new net.minecraft.network.chat.TextComponent("");
```
Comment (tahmid-23) [inline patches/server/0803-Implement-server-resource-pack-API.patch]: does this not work
```suggestion
+            net.minecraft.network.chat.TextComponent.EMPTY;
```
also, why use an empty component and not just null?

File: patches/server/0803-Implement-server-resource-pack-API.patch [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -88,8 +88,12 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+     private MinecraftServerGui gui;
+     @Nullable
+     private final TextFilterClient textFilterClient;
++    // Paper start
+     @Nullable
+-    private final Component resourcePackPrompt;
++    private Component resourcePackPrompt;
```
Comment (tahmid-23) [inline patches/server/0803-Implement-server-resource-pack-API.patch]: I believe this can use an access transformer
build-data/paper.at

**Comment by Machine-Maker** [comment]
Something about this feels weird, I can't put my finger on it though. I know you can already send resource packs to different players, but I don't know how this would interact with that. Can you only have 1 resource pack from a server applied at a time?

File: patches/server/0803-Implement-server-resource-pack-API.patch [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
++    @Nullable
```
Comment (tahmid-23) [inline patches/server/0803-Implement-server-resource-pack-API.patch]: why make this nullable? from what I see you only set it to a `boolean` value and if it is null you use the server properties value, so why not just read the server property upon construction?

File: patches/server/0803-Implement-server-resource-pack-API.patch (hunk 2 above)
Comment (yusshu) [reply patches/server/0803-Implement-server-resource-pack-API.patch]: Didn't know about that, done

File: patches/server/0803-Implement-server-resource-pack-API.patch (hunk 3 above)
Comment (yusshu) [reply patches/server/0803-Implement-server-resource-pack-API.patch]: done

File: patches/server/0803-Implement-server-resource-pack-API.patch (hunk 1 above)
Comment (yusshu) [reply patches/server/0803-Implement-server-resource-pack-API.patch]: Done, using null

File: patches/server/0803-Implement-server-resource-pack-API.patch [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -88,8 +88,11 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+     private MinecraftServerGui gui;
+     @Nullable
+     private final TextFilterClient textFilterClient;
++    // Paper start
+     @Nullable
+     private Component resourcePackPrompt;
++    private boolean resourcePackRequired;
++    // Paper end
```
Comment (tahmid-23) [inline patches/server/0803-Implement-server-resource-pack-API.patch]: remove the Paper start and Paper end and just make it a single line comment

**Comment by iAtog** [comment]
good pr

**Comment by Owen1212055** [comment]
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/439c685f4ea3452ce07adad6f4da7526daac2eac#src/main/java/org/bukkit/craftbukkit/CraftServer.java

Do note, seems to have been halfway implemented upstream. It allows you to get the resource pack, but not to set it.

**Pull request closed without merging**